Document SAP Fiori 3 Styling (#2857)
mendix/docs: content/partners/sap/sap-fiori-3-0.md

The examples of Fiori 3 styling in the SAP Fiori 3 UI Resources App Store Module will be clearer if you populate the database first using the **Create Test Data** option on the Home page.

### 2.1 Examples of Mendix Core Widget Styling

You can see examples of the Fiori 3 styling of the Mendix Core Widgets by going to the **Mendix_Core_Elements** page of the SAP Fiori 3 UI Resources App Store Module.

Most widgets work the same way as they do in other Mendix styles based on Atlas UI, but there are some changes which are listed in the sections below.

#### 2.1.1 Buttons

You can choose different types of button by setting the appropriate **Button style**. The SAP Fiori styles are mapped as follows:

| **SAP Fiori Style** | **Mendix Button Style** |
| --- | --- |
| Default | Default |
| Emphasized | Primary |
| Positive | Success |
| Negative | Danger |
| Attention | Warning |
| Ghost | *requires manual CSS class, see [Applying CSS Classes Manually](#apply-css), below* |

{{% image_container width="75%" %}}
![SAP Fiori Button Styles](attachments/sap-fiori-3-0/button-styles.png)
{{% /image_container %}}

The **Link button** also has Fiori 3 styling.

#### 2.1.2 Action Bar

> {{% todo %}}[Does the construction of the Action Bar need to be described? Why isn't it a building block?]{{% /todo %}}

Reviewer: Not required. Let's remove this from mendix core elements.

The **container count badge** allows you to add a count badge to containers on Mendix pages as shown in the following example.

![SAP Fiori Count Badge Example](attachments/sap-fiori-3-0/count-badge-example.png)

You can create this style by adding the **ContainerCountBadge** widget inside a container.

![SAP Fiori Count Badge page design](attachments/sap-fiori-3-0/count-badge-edit.png)

The number which is displayed in the badge is supplied by a microflow which you can specify in the **Data Source** tab of the badge properties. The microflow must return an integer value.

![SAP Fiori Count Badge microflow](attachments/sap-fiori-3-0/count-badge-microflow.png)

You can specify the appearance of the count badge in the **Appearance** tab of the badge properties. If you want to maintain the correct SAP Fiori theme, you should choose the **Color Source** *Bootstrap Color Schema*.

![SAP Fiori Count Badge appearance](attachments/sap-fiori-3-0/count-badge-appearance.png)

#### 2.1.4 List Views

Standard Mendix **List view** widgets can be used on your pages. However, these cannot be borderless. To create a borderless list view, see [Applying CSS Classes Manually](#apply-css), below.

> {{% todo %}}[Do we need a section }#### 2.1.5 Pagination]{{% /todo %}}

Reviewer: Actually not, because Adding a pagination is same as of Atlas UI, so if we can assume that develop is aware of developing app in Mendix,

### 2.2 Applying CSS Classes Manually{#apply-css}

Some SAP Fiori styling requires you to add some additional CSS classes manually for the styles to be implemented correctly. These cases are described in the following sections.

#### 2.2.1 Transparent Buttons

#### 2.2.2 Borderless List Views

### 2.3 Unsupported SAP Fiori Components{#unsupported-components}

* Avatar – you can mimic this using a **List 3** building block
* Message Strip – we suggest you use one of the **Notifications** building blocks 
* Message Toast/Short message on actions

> * Time

Reviewer: Added - Timepicker control not supported

* StepInput

> * Menu/Navigation

Reviewer: Not supported in menus
Menu having two icons. 
Menu having slider right to show submenu

> * Calendar

Reviewer: Calendar highlighting multiple Dates like Holidays, any other special days can't be supported

> * Tabs

Reviewer: Icons and counter can’t be supported
Icons only Tab can’t be supported.

> * Table

Reviewer: Table having checkboxes and icons can’t be supported. 

Table row selection is possible at row selection level.

Author: I've added the bit about row selection, but I don't really understand what it means.

Reviewer: SAP Fiori allows to show checkboxes to select a row, that is not possible mendix. In order to select row/rows from a table, "selection Mode" properties is available in Table properties.

Author: Updated this.

## 3 Building Blocks{#building-blocks}

You can see examples of the Fiori 3 building blocks by going to the **Building Blocks** page of the SAP Fiori 3 UI Resources App Store Module. The building blocks are categorized by type.

> {{% todo %}}[Is there anything that needs to be mentioned here? Are there any specific use cases or configuration (e.g. for notifications)]{{% /todo %}}

Reviewer: Didn't get this point?

Author: I just wanted to know if there was anything specific we needed to write about the building blocks. For example, how you might use the _Flex Container Left Left_, what the _Shell Bar_ is, how you might use a _Notification Action_ or what the difference is between _List 1_ and _List 4_

Reviewer: we can add that for few of the building blocks. Let's keep this as improvement to this section. First release this and then iteratively we can add information to it.

Author: OK - I'll mark this resolved and we can work on the next iteration.

* Use a *Sliding* layout to display dynamic side content. See [Dynamic Side Content](https://experience.sap.com/fiori-design-web/dynamic-side-content/) on the *SAP Fiori Design Guidelines* site for more information.
* Use a *Letterbox* layout use Letterboxing display to restrict your layout to a certain width. See [Letterboxing](https://experience.sap.com/fiori-design-web/letter-boxing/) on the *SAP Fiori Design Guidelines* site for more information.
* Use an *Embedded* layout when you are using this page inside a page which already contains a Header.

Here is a list of the layouts you can use.

* PopupLayout (SAP_3_0_UI_Resources) 
* SAP_MasterDetail_Embedded (SAP_3_0_UI_Resources) 
* SAP_Launchpad (SAP_3_0_UI_Resources) 
* SAP_Default_Letterbox (SAP_3_0_UI_Resources) 
* SAP_MasterDetail_Sliding (SAP_3_0_UI_Resources)
* SAP_MasterDetail_Sliding_Letterbox (SAP_3_0_UI_Resources)
* SAP_Default_Embedded (SAP_3_0_UI_Resources)
* SAP Default (SAP_3_0_UI_Resources)
* SAP_MasterDetail_Letterbox (SAP_3_0_UI_Resources)
* SAP_MasterDetail (SAP_3_0_UI_Resources) 

## 5 Page Templates{#page-templates}

> There are many different page layouts which you can use to base your page on. These provide you with the building blocks and widgets already arranged to provide an SAP Fiori-style page.

Reviewer: different page templates

Note that these pages work best when combined with the appropriate **Navigation layout**. For example, the **Fiori Launchpad** page template works best with the *SAP_Launchpad (SAP_3_0_UI_Resources)* **Navigation Layout**.

## 6 Theme Selector{#theme-selector}

You can give the end-use the option to choose their own theme. There is an example of this on the **Home** page of the SAP Fiori 3 UI Resources App Store Module. The end-user can click a button to switch between a *Dark Theme* or a *Light Theme*.

To do this you need to trigger a nanoflow which passes a variable containing the name of the theme to the **ChangeTheme_JavaScript_Action** JavaScript action. This changes the theme and implements it directly on your app's pages.

The themes available are:

> * **light** *(default)*

Reviewer: Quartz light

Author: The values I put here are the ones which are used in the Java code and need to be passed in the Java action
![image](https://user-images.githubusercontent.com/35492184/99389115-d8c78980-28d6-11eb-823e-a0b594198d8d.png)
![image](https://user-images.githubusercontent.com/35492184/99389254-0ad8eb80-28d7-11eb-9769-c8fbfb66f985.png)

> * **dark**

Reviewer: Quartz Dark

Author: Not the value that the Java Code takes.

Reviewer: yes you are correct. the name mentioned here are the theme names in our SAP Fiori library. What I want to mentioned that theme-light is corresponds to SAP Quartz light and "theme-dark" is correspond to "SAP Quartz Dark".

Author: OK - I'll add this correspondence.

---
title: "SAP Fiori 3.0 Styling"
category: "SAP"
menu_order: 70
description: "Information on how to make an SAP Fiori version 3.0 themed app"
tags: ["SAP", "Fiori", "Building Block", "Page Design", "Layout", "UI", "UX", "3.0"]
---

<!-- NOTES
Check all links
Should we have a section on the "Shell Bar"?

-->

## 1 Introduction

> SAP Fiori 3 is SAP's target design system used to give your SAP C4/HANA application a consistent look and feel. The [SAP Fiori 3 UI Resources](#wrong) module in the Mendix App Store allows you to apply this look and feel to your Mendix apps running on SAP C4/HANA.

Reviewer 2: SAP Fiori user experience is a new visual design and appearance for creating UI across SAP Suite of products. SAP Fiori 3 is SAP’s new target design system for all SAP products to support the Intelligent Suite. With [SAP Fiori 3 UI Resource] module within Mendix AppStore, we bring the same visual experience for users to create apps using SAP Fiori 3 design language

Author: @satyammx I feel we are doing SAP's selling work here - this is a technical document describing how we are supporting SAP Fiori if the user has decided to use it, not trying persuade people to switch to SAP Fiori.

```suggestion
The SAP Fiori user experience creates a consistent UI across the SAP Suite of products. SAP Fiori 3 is SAP’s new target design system for all SAP products in the *Intelligent Suite*. The  [SAP Fiori 3 UI Resource](#wrong) module in the Mendix App Store allows you to create the same user experience in your Mendix apps.
```

Reviewer 2: I like this overview much better. 
"SAP Selling" - we discussed that on slack :)

The Mendix module gives you access to designing your pages with SAP Fiori-styled widgets and building blocks as part of your layout. You can also allow end-users to select a theme and apply it immediately. The module supports SAP Quartz Light and Dark themes.

You can experience the SAP Fiori 3 design by using the [SAP Fiori 3 Starter App](#wrong) as the basis for a new app, or by including the [SAP Fiori 3 UI Resources](#wrong) module in your app.

More information on using SAP Fiori 3 design in your Mendix app is available in the following sections:

> * [Widgets](#widgets) – Mendix core widgets have SAP Fiori 3 styling

Reviewer 2: Mendix core widgets having SAP Fiori 3 styling

Author: ```suggestion
* [Widgets](#widgets) – Mendix core widgets having SAP Fiori 3 styling
```

> * [Building Blocks](#building-blocks) – groups of widgets which can be used to quickly create SAP Fiori styled pages

Reviewer 2: instead of 
groups of widgets
we can use
grouped widgets

Author: I'm not sure that "grouped widgets" is more accurate. Is there any reason you want to change it?

Reviewer 2: You are right. on re-reading this, grouped widgets seems less accurate

* [Layouts](#layouts) – different styles of page which can be used in your app — for example, Master/detail or Pop-Up
* [Page Templates](#page-templates) – readily-available page templates which are already styled ready for you to use in your Mendix app
* [Theme Selector](#theme-selector) – switch between different theme — for example *light* and *dark*

## 2 Widgets{#widgets}

The majority of Mendix widgets have been styled to match SAP Fiori 3 styles by default. Examples are shown below.

> There are a few cases where you will need to apply some extra CSS classes manually to achieve the correct SAP Fiori look and feel. These are described in [Applying CSS Classes Manually](#apply-css).

Reviewer 2: instead of 
There are a few cases where you will need to apply ... ...
may be
In exceptional cases, you may need to apply ... ...

Author: Is wanting to use a disabled link or an inverted link exceptional?
Perhaps
```suggestion
There are a couple of cases where you will need to apply some extra CSS classes manually to achieve the correct SAP Fiori look and feel. These are described in [Applying CSS Classes Manually](#apply-css).
```

Reviewer 2: Looks ok to me.

> There are also a few SAP Fiori components which are not supported by Mendix core widgets. These are listed in [Unsupported SAP Fiori Components](#unsupported-components), below.

Reviewer 2: instead of 
There are also a few SAP Fiori components which are not supported by Mendix core widgets
may be
Some Fiori 3 components are not yet supported by Mendix core widgets.

Author: I assume this means we are planning to support them in a future release?
```suggestion
Some SAP Fiori components are not yet supported by Mendix core widgets. These are listed in [Unsupported SAP Fiori Components](#unsupported-components), below.
```

| Emphasized Link | link-emphasized |
| Inverted Link | link-inverted |
| Right Icon Link | link right icon |
| Left Icon Link | *this is the default* |

### 2.3 Unsupported SAP Fiori Components{#unsupported-components}

* Avatar – you can mimic this using a **List 3** building block
* Message Strip – we suggest you use one of the **Notifications** building blocks 
* Message Toast/Short message on actions
* Time
* Timepicker
* StepInput

In addition, there are some limitations on supporting other components:

* In **Menus** you cannot have:
    * two icons
    * a slider showing sub-menus
* In **Calendars** you cannot:
    * highlight multiple dates such as holidays or other special days
* In **Tables**
    * you cannot use icons on tables
    * you cannot select rows using checkboxes, you can only select rows using the standard Mendix selection methods
* In **Tabs** you cannot have
    * icons with a counter
    * only icons

## 3 Building Blocks{#building-blocks}

You can see examples of the Fiori 3 building blocks by going to the **Building Blocks** page of the SAP Fiori 3 UI Resources App Store Module. The building blocks are categorized by type.

## 4 Layouts{#layouts}

> You can see examples of the Fiori 3 navigation layouts by going to the **Layout** page of the SAP Fiori 3 UI Resources App Store Module. These layouts are designed for use in *Responsive (Web)* applications, designed to run in a browser.

Reviewer 2: instead of 
designed to run in a browser
may be
designed to run on any modern web browser, across any device

Author: ```suggestion
You can see examples of the Fiori 3 navigation layouts by going to the **Layout** page of the SAP Fiori 3 UI Resources App Store Module. These layouts are designed for use in *Responsive (Web)* applications, designed to run in any modern web browser, on any device.
```

* Use a *Default* layout to display information in a single panel.
* Use a *MasterDetail* layout to display information in two panels side-by-side such as an *Order* with multiple *Order Lines*.
* Use a *Sliding* layout to display dynamic side content. See [Dynamic Side Content](https://experience.sap.com/fiori-design-web/dynamic-side-content/) on the *SAP Fiori Design Guidelines* site for more information.
* Use a *Letterbox* layout use Letterboxing display to restrict your layout to a certain width. See [Letterboxing](https://experience.sap.com/fiori-design-web/letter-boxing/) on the *SAP Fiori Design Guidelines* site for more information.
* Use an *Embedded* layout when you are using this page inside a page which already contains a Header.

Here is a list of the layouts you can use.

* PopupLayout (SAP_3_0_UI_Resources) 
* SAP_MasterDetail_Embedded (SAP_3_0_UI_Resources) 
* SAP_Launchpad (SAP_3_0_UI_Resources) 
* SAP_Default_Letterbox (SAP_3_0_UI_Resources) 
* SAP_MasterDetail_Sliding (SAP_3_0_UI_Resources)
* SAP_MasterDetail_Sliding_Letterbox (SAP_3_0_UI_Resources)
* SAP_Default_Embedded (SAP_3_0_UI_Resources)

> * SAP Default (SAP_3_0_UI_Resources)

Reviewer 2: is underscore missing? 
SAP_Default

Author: Thanks - well spotted.
```suggestion
* SAP_Default (SAP_3_0_UI_Resources)
```

* SAP_MasterDetail_Letterbox (SAP_3_0_UI_Resources)
* SAP_MasterDetail (SAP_3_0_UI_Resources) 

## 5 Page Templates{#page-templates}

> There are many different page templates which you can use to base your page on. These provide you with the building blocks and widgets already arranged to provide an SAP Fiori-style page.

Reviewer 2: Instead of 
There are many different page templates which you can use to base your page on. These provide you ...
may be
You may use one of the available page templates to rapidly create a compelling UI for your app. The available templates provide you ...

Author: I find this is selling Fiori again.
How about
```suggestion
You can use one of the available page templates to rapidly create a consistent UI for your app pages. The templates provide you with the building blocks and widgets already arranged to provide an SAP Fiori-style page.
```

Reviewer 2: This is much better in my viewpoint.

Note that these pages work best when combined with the appropriate **Navigation layout**. For example, the **Fiori Launchpad** page template works best with the *SAP_Launchpad (SAP_3_0_UI_Resources)* **Navigation Layout**.

## 6 Theme Selector{#theme-selector}

You can give the end-use the option to choose their own theme. There is an example of this on the **Home** page of the SAP Fiori 3 UI Resources App Store Module. The end-user can click a button to switch between a *Dark Theme* or a *Light Theme*.

To do this you need to trigger a nanoflow which passes a variable containing the name of the theme to the **ChangeTheme_JavaScript_Action** JavaScript action. This changes the theme and implements it directly on your app's pages.

The themes available are:

> * **light** *(default)* – the SAP Quartz Light theme

Reviewer 2: instead of 
the SAP Quartz Light theme
may be
equivalent to the SAP Quartz Light theme

Author: ```suggestion
* **light** *(default)* – equivalent to the SAP Quartz Light theme
```

> * **dark** – the SAP Quartz Dark theme

Reviewer 2: similar comment as above

> * **dark** – the SAP Quartz Dark theme

Reviewer 2: instead of
the SAP Quartz Dark theme
may be
equivalent to the SAP Quartz Dark theme

Author: ```suggestion
* **dark** – equivalent to the SAP Quartz Dark theme
```

The majority of Mendix widgets have been styled to match SAP Fiori 3 styles by default. Examples are shown below.

There are a few cases where you will need to apply some extra CSS classes manually to achieve the correct SAP Fiori look and feel. These are described in [Applying CSS Classes Manually](#apply-css).

There are also a few SAP Fiori components which are not supported by Mendix core widgets. These are listed in [Unsupported SAP Fiori Components](#unsupported-components), below.

### 2.1 Examples of Mendix Core Widget Styling

You can see examples of the Fiori 3 styling of the Mendix Core Widgets by going to the **Mendix_Core_Elements** page of the SAP Fiori 3 UI Resources App Store Module.

Most widgets work the same way as they do in other Mendix styles based on Atlas UI, but there are some changes which are listed in the sections below.

#### 2.1.1 Buttons

You can choose different types of button by setting the appropriate **Button style**. The SAP Fiori styles are mapped as follows:

| **SAP Fiori Style** | **Mendix Button Style** |
| --- | --- |
| Default | Default |
| Emphasized | Primary |
| Positive | Success |
| Negative | Danger |
| Attention | Warning |

> | Ghost | *requires manual CSS class, see [Applying CSS Classes Manually](#apply-css), below* |

Reviewer: Replace ghost with "Transparent". I have updated it in app as well.

{{% image_container width="75%" %}}
![SAP Fiori Button Styles](attachments/sap-fiori-3-0/button-styles.png)
{{% /image_container %}}

> The **Link button** also has Fiori 3 styling.

Reviewer: change it to for Link Button, see [Applying CSS Classes Manually](#apply-css), below* |

#### 2.1.2 Container Count Badge

The **container count badge** allows you to add a count badge to containers on Mendix pages as shown in the following example.

![SAP Fiori Count Badge Example](attachments/sap-fiori-3-0/count-badge-example.png)

You can create this style by adding the **ContainerCountBadge** widget inside a container.

![SAP Fiori Count Badge page design](attachments/sap-fiori-3-0/count-badge-edit.png)

The number which is displayed in the badge is supplied by a microflow which you can specify in the **Data Source** tab of the badge properties. The microflow must return an integer value.

![SAP Fiori Count Badge microflow](attachments/sap-fiori-3-0/count-badge-microflow.png)

You can specify the appearance of the count badge in the **Appearance** tab of the badge properties. If you want to maintain the correct SAP Fiori theme, you should choose the **Color Source** *Bootstrap Color Schema*.

![SAP Fiori Count Badge appearance](attachments/sap-fiori-3-0/count-badge-appearance.png)

#### 2.1.3 List Views

> Standard Mendix **List view** widgets can be used on your pages. However, these cannot be borderless. To create a borderless list view, see [Applying CSS Classes Manually](#apply-css), below.

Reviewer: remove the line "To create a borderless list view, see [Applying CSS Classes Manually](#apply-css), below."
Because borderless list is available as building block.

### 2.2 Applying CSS Classes Manually{#apply-css}

Some SAP Fiori styling requires you to add some additional CSS classes manually for the styles to be implemented correctly. These cases are described in the following tables.

| SAP Component Style | Mendix Class(es) |
| --- | --- |
| Disabled Link | link-disabled |
| Emphasized Link | link-emphasized |
| Inverted Link | link-inverted |
| Right Icon Link | link right icon |
| Left Icon Link | *this is the default* |

> | Transparent Button | transparent-btn |

Author: Do we need a class for ** borderless list views** ?

Reviewer: No not required it is added as building block
